### ISSUE TITLE
TypeVar constrained to overlapping Union types does not accept common type

### ISSUE THREAD

**Issue by FynnBe** [microsoft/pyright#7398]
In the code example below `a` complies to both union types `U1` and `U2` to which TypeVar `T` is bound, but an error is raised on its assignment regardless.

```python
from typing import Literal, TypeVar, Union, reveal_type

U1 = Union[Literal["A"], Literal[1]]
U2 = Union[Literal["A"], Literal[2]]

T = TypeVar("T", U1, U2)


def foo(bar: T, a: Literal["A"] = "A") -> list[T]:
    ret: list[T] = [bar]
    ret.append(a)  # reportArgumentType error
    return ret


result1 = foo(1)  # list[U1] as expected
result1.append("A")  # no error
print(reveal_type(result1))

result2 = foo(2)  # list[U2] as expected
result2.append("A")  # no error
print(reveal_type(result2))

```

- Python 3.12.1
- pyright 1.1.352

**Comment by erictraut** [comment]
This won't work given the way that [pyright implements constrained TypeVars](https://microsoft.github.io/pyright/#/mypy-comparison?id=value-constrained-type-variables). 

Constrained TypeVars are an oddity of the Python type system, not found in any other language. I'm not sure why you're attempting to use them here, but I recommend looking for an alternative.
